### ISSUE TITLE
Nix should ignore or invalidate broken store paths in cache

### ISSUE THREAD

**Issue by thomasjm** [NixOS/nix#2519]
Filing this after discussion on #nixos channel. I managed to get the Nix cache in `~/.cache/nix` in a bad state via the following steps:

1. Install Nix normally via `curl https://nixos.org/nix/install | sh`
2. Decide I want to have a custom store path. Delete `/nix` and `~/.nix*` and assume Nix was totally removed from the system. (It was not, `~/.cache/nix` still existed.)
3. Download Nix source and configure with `--with-store-dir=/nix_frozen/store --localstatedir=/nix_frozen/var`, make and install.
4. Activate profile with `[prefix]/etc/profile.d/nix.sh`
5. Install a random package with `nix-env -f ~/tools/nixpkgs awscli` to get profiles created.

At this point, `nix-channel --update` fails with the error `error: path '/nix/store/p9bczv4dp4yk8qp9pq5sdvh56pl3hxyd-nixpkgs-unstable' is not in the Nix store`. This is because there was a broken symlink to this path in the cache from the previous installation.

The desired behavior would be to detect broken symlinks in the cache and either ignore or remove them.

**Comment by domenkozar** [comment]
cc @Infinisil I think that's the one you hit a couple of days ago

**Comment by Infinisil** [comment]
Yeah something similar happened to me, no idea how to fix it though

**Comment by thomasjm** [comment]
I've been doing some more work where I have two Nix setups installed side by side, one with prefix `/nix` and one with `/nix_frozen`. These cache collisions are even more of a problem if you try to switch back and forth between them like this.

Having thought about it a bit more, I think the right solution is actually for the `~/.cache/nix` folder to simply be keyed by the Nix prefix. So instead of `~/.cache/nix/[cache stuff]`, you have `~/.cache/nix/[prefix]/[cache stuff]`. 

This way I could have `~/.cache/nix/nix/[stuff]` and `~/.cache/nix/nix_frozen/[stuff]` side by side on my system.

(Or alternatively perhaps, use the value of `NIX_STORE_DIR` as the key.)

Does this sound reasonable?

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
